File: src/bun/core/rank.ts
const DIGITS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
const MIN_DIGIT = 0;
const MAX_DIGIT = DIGITS.length - 1;
const MID_DIGIT = Math.floor(MAX_DIGIT / 2);

function digitAt(value: string | null, index: number, fallback: number) {
	if (value == null || index >= value.length) return fallback;
	const digit = DIGITS.indexOf(value[index]);
	if (digit === -1) {
		throw new Error(`Invalid rank character: ${value[index]}`);
	}
	return digit;
}

export function rankBetween(before: string | null, after: string | null) {
	if (before != null && after != null && before >= after) {
		throw new Error(`Cannot rank between ${before} and ${after}`);
	}

	let prefix = "";

	for (let index = 0; ; index += 1) {
		const low = digitAt(before, index, MIN_DIGIT);
		const high = digitAt(after, index, MAX_DIGIT);

		if (high - low > 1) {
			return `${prefix}${DIGITS[Math.floor((low + high) / 2)]}`;
		}

		prefix += DIGITS[low] ?? DIGITS[MID_DIGIT];
	}
}

File: src-tauri/src/core/rank.rs
use super::Result;

const DIGITS: &str = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

pub(crate) fn rank_between(before: Option<&str>, after: Option<&str>) -> Result<String> {
    if let (Some(before), Some(after)) = (before, after) {
        if before >= after {
            return Err(format!("Cannot rank between {before} and {after}"));
        }
    }

    let digits: Vec<char> = DIGITS.chars().collect();
    let min_digit = 0;
    let max_digit = digits.len() - 1;
    let mid_digit = max_digit / 2;
    let mut prefix = String::new();

    for index in 0.. {
        let low = digit_at(before, index, min_digit)?;
        let high = digit_at(after, index, max_digit)?;

        if high - low > 1 {
            prefix.push(digits[(low + high) / 2]);
            return Ok(prefix);
        }

        prefix.push(*digits.get(low).unwrap_or(&digits[mid_digit]));
    }

    unreachable!()
}

fn digit_at(value: Option<&str>, index: usize, fallback: usize) -> Result<usize> {
    let Some(value) = value else {
        return Ok(fallback);
    };
    let Some(character) = value.chars().nth(index) else {
        return Ok(fallback);
    };
    DIGITS
        .chars()
        .position(|digit| digit == character)
        .ok_or_else(|| format!("Invalid rank character: {character}"))
}
